code/go/0chain.net/blobbercore/filestore/file_manager.go

```diff
@@ -0,0 +1,495 @@
+//
+// Above situation also occurs to store {allocation_id} as base directory for its files when blobber can have thousands of allocations.
+// We can also use allocation_id to distribute allocations.
+// For allocation using 3 levels we would have 16*16*16 = 4096 unique directories, Each directory can contain 1000 allocations. Thus storing
+// 4096000 number of allocations.
+//
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/0chain/blobber/code/go/0chain.net/blobbercore/datastore"
+	"github.com/0chain/blobber/code/go/0chain.net/core/common"
+	"github.com/0chain/blobber/code/go/0chain.net/core/logging"
+	"go.uber.org/zap"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	MaxFilesInADir   = 1000
+	SmallestFileSize = 64 * 1024
+	TempDir          = "tmp"
+)
+
+var currentDiskCapacity uint64
+var getMountPoint func() string
+var isMountPoint func(mp string) bool = func(mp string) bool {
+	if !filepath.IsAbs(mp) {
+		logging.Logger.Error(fmt.Sprintf("%s is not absolute path", mp))
+		return false
+	}
+
+	realMP, err := filepath.EvalSymlinks(mp)
+	if err != nil {
+		logging.Logger.Error(err.Error())
+		return false
+	}
+
+	finfo, err := os.Lstat(realMP)
+	if err != nil {
+		logging.Logger.Error(err.Error())
+		return false
+	}
+
+	pinfo, err := os.Lstat(filepath.Dir(realMP))
+	if err != nil {
+		logging.Logger.Error(err.Error())
+		return false
+	}
+
+	dev := finfo.Sys().(*syscall.Stat_t).Dev
+	pDev := pinfo.Sys().(*syscall.Stat_t).Dev
+
+	return dev != pDev
+}
+
+type allocation struct {
+	mu            *sync.Mutex
+	allocatedSize uint64
+	filesNumber   uint64
+	filesSize     uint64
+
+	tmpMU       *sync.Mutex
+	tmpFileSize uint64
+}
+
+type fileManager struct {
+	// allocMu is used to update especially add new allocation object
+	allocMu     *sync.Mutex
+	rwMU        *sync.RWMutex
+	Allocations map[string]*allocation
+}
+
+// UpdateAllocationMetaData only updates if allocation size has changed or new allocation is allocated. Must use allocationID.
+// Use of allocation Tx might leak memory. allocation size must be of int64 type otherwise it won't be updated
+func UpdateAllocationMetaData(m map[string]interface{}) {
+	fm.allocMu.Lock()
+	defer fm.allocMu.Unlock()
+
+	allocIDI := m["allocation_id"]
+	if allocIDI == nil {
+		return
+	}
+
+	allocID, ok := allocIDI.(string)
+	if !ok {
+		return
+	}
+
+	allocatedSizeI := m["allocated_size"]
+	if allocatedSizeI == nil {
+		return
+	}
+
+	allocatedSize, ok := allocatedSizeI.(int64)
+	if !ok {
+		return
+	}
+	alloc := fm.getAllocation(allocID)
+	if alloc == nil {
+		alloc = &allocation{
+			allocatedSize: uint64(allocatedSize),
+			mu:            &sync.Mutex{},
+			tmpMU:         &sync.Mutex{},
+		}
+
+		fm.setAllocation(allocID, alloc)
+		return
+	}
+
+	alloc.allocatedSize = uint64(allocatedSize)
+
+}
+
+func (fm *fileManager) getAllocation(allocID string) *allocation {
+	fm.rwMU.RLock()
+	defer fm.rwMU.Unlock()
```

Comment:
Don't understand why. Can you please explain why it should not be used?

Reply:
Defer has some performance consideration, and for this three line of code, it could be done in the following way without defer and having reasonable readability. Anyway, it's not a big problem, but I would recommend to do it without defer.

```go
func (fm *fileManager) getAllocation(allocID string) (alloc *allocation) {
    fm.rwMu.RLock()
    alloc = fs.mAllocs[allocID]
    fm.rwMu.RUnlock()
    return
}
```